policy/score_calculator_test.go

```diff
@@ -404,3 +405,129 @@ func TestImpact(t *testing.T) {
+				Value: int32(impact),
+			},
+		})
+	}
+}
+
+func addTestResults(critFail, critPass, highFail, highPass, midFail, midPass, lowFail, lowPass int, calculator ScoreCalculator) {
+	addMultipleScores(100, critFail, critPass, calculator)
+	addMultipleScores(80, highFail, highPass, calculator)
+	addMultipleScores(55, midFail, midPass, calculator)
+	addMultipleScores(20, lowFail, lowPass, calculator)
+}
+
+func testEachScoreCalculator(critFail, critCnt, highFail, highCnt, midFail, midCnt, lowFail, lowCnt int, t *testing.T) []*Score {
+	require.LessOrEqual(t, critFail, critCnt)
+	require.LessOrEqual(t, highFail, highCnt)
+	require.LessOrEqual(t, midFail, midCnt)
+	require.LessOrEqual(t, lowFail, lowCnt)
+
+	fmt.Printf("Results: %d/%d CRIT   %d/%d HIGH   %d/%d MID   %d/%d LOW   %d/%d Total\n",
+		critFail, critCnt,
+		highFail, highCnt,
+		midFail, midCnt,
+		lowFail, lowCnt,
+		critFail+highFail+midFail+lowFail, critCnt+highCnt+midCnt+lowCnt,
+	)
+	calculators := []ScoreCalculator{
+		&averageScoreCalculator{},
+		&bandedScoreCalculator{},
+		&decayedScoreCalculator{},
+		&worstScoreCalculator{},
+	}
+
+	var results []*Score
+	for i := range calculators {
+		calculator := calculators[i]
+		calculator.Init()
+
+		addTestResults(critFail, critCnt-critFail, highFail, highCnt-highFail, midFail, midCnt-midFail, lowFail, lowCnt-lowFail, calculator)
+
+		score := calculator.Calculate()
+		results = append(results, score)
+		require.NotNil(t, score)
+		fmt.Printf("    %20s: %3d  (%s)\n", calculator.String(), score.Value, score.Rating().FailureLabel())
+	}
+	fmt.Println("")
+
+	return results
+}
+
+func TestCalculatorBehavior_Banded(t *testing.T) {
+	var res []*Score
+	t.Run("critical band", func(t *testing.T) {
+		res = testEachScoreCalculator(1, 10, 0, 10, 0, 10, 0, 10, t)
+		assert.Equal(t, 45, int(res[1].Value))
+		res = testEachScoreCalculator(5, 10, 0, 10, 0, 10, 0, 10, t)
+		assert.Equal(t, 25, int(res[1].Value))
+		res = testEachScoreCalculator(10, 10, 0, 10, 0, 10, 0, 10, t)
+		assert.Equal(t, 0, int(res[1].Value))
+	})
+	t.Run("critical band with high band variation", func(t *testing.T) {
+		res = testEachScoreCalculator(1, 10, 1, 10, 0, 10, 0, 10, t)
+		assert.Equal(t, 41, int(res[1].Value))
+		res = testEachScoreCalculator(1, 10, 5, 10, 0, 10, 0, 10, t)
+		assert.Equal(t, 27, int(res[1].Value))
+		res = testEachScoreCalculator(1, 10, 10, 10, 0, 10, 0, 10, t)
+		assert.Equal(t, 9, int(res[1].Value))
+	})
+	t.Run("high band", func(t *testing.T) {
+		res = testEachScoreCalculator(0, 10, 1, 10, 0, 10, 0, 10, t)
+		assert.Equal(t, 55, int(res[1].Value))
+		res = testEachScoreCalculator(0, 10, 5, 10, 0, 10, 0, 10, t)
+		assert.Equal(t, 35, int(res[1].Value))
+		res = testEachScoreCalculator(0, 10, 10, 10, 0, 10, 0, 10, t)
+		assert.Equal(t, 10, int(res[1].Value))
+	})
+}
+
+// This is a toolchain that can be used to see what the scoring mechanism would produce under different scenarios
+func TestScoreMechanisms(t *testing.T) {
```

Comment:
This has been added to have a playground for experimenting with scoring mechanisms. Uncomment the last line to force the failure to be printed